sycl/include/sycl/ext/oneapi/annotated_arg/properties.hpp

```diff
@@ -0,0 +1,307 @@
+// properties.hpp - SYCL properties associated with annotated_arg/ptr //
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/ext/oneapi/properties/property.hpp>
+#include <sycl/ext/oneapi/properties/property_value.hpp>
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+
+template <typename T, typename PropertyListT> class annotated_arg;
+
+//===----------------------------------------------------------------------===//
+//        Common properties of annotated_arg/annotated_ptr
+//===----------------------------------------------------------------------===//
+struct register_map_key {
+  using value_t = property_value<register_map_key>;
+};
+
+struct conduit_key {
+  using value_t = property_value<conduit_key>;
+};
+
+struct stable_key {
+  using value_t = property_value<stable_key>;
+};
+
+struct buffer_location_key {
+  template <int K>
+  using value_t =
+      property_value<buffer_location_key, std::integral_constant<int, K>>;
+};
+
+struct awidth_key {
+  template <int K>
+  using value_t = property_value<awidth_key, std::integral_constant<int, K>>;
+};
+
+struct dwidth_key {
+  template <int K>
+  using value_t = property_value<dwidth_key, std::integral_constant<int, K>>;
+};
+
+struct latency_key {
+  template <int K>
+  using value_t = property_value<latency_key, std::integral_constant<int, K>>;
+};
+
+enum class read_write_mode_enum : std::uint16_t { read_write, read, write };
+
+struct read_write_mode_key {
+  template <read_write_mode_enum Mode>
+  using value_t =
+      property_value<read_write_mode_key,
+                     std::integral_constant<read_write_mode_enum, Mode>>;
+};
+
+struct maxburst_key {
+  template <int K>
+  using value_t = property_value<maxburst_key, std::integral_constant<int, K>>;
+};
+
+struct wait_request_key {
+  template <int K>
+  using value_t =
+      property_value<wait_request_key, std::integral_constant<int, K>>;
+};
+
+#if __cplusplus >= 201703L // inline variables
```

Comment:
C++17 is a required minimum version now to compile SYCL headers, you can omit the check. See, for example #7687